File: app/src/main/java/com/anthonymandra/rawdroid/data/MetadataEntity.java
package com.anthonymandra.rawdroid.data;

import android.arch.persistence.room.Entity;
import android.arch.persistence.room.ForeignKey;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.Index;
import android.arch.persistence.room.PrimaryKey;

import static android.arch.persistence.room.ForeignKey.CASCADE;


@Entity(tableName = "meta",
		foreignKeys = @ForeignKey(entity = FolderEntity.class,
				parentColumns = "id",
				childColumns = "parentId",
				onDelete = CASCADE),
		indices = { @Index(value = "uri", unique = true),
					@Index(value = "documentId", unique = true),
					@Index(value = "parentId")})
public class MetadataEntity
{
	public static MetadataEntity createMetadataEntity(SearchEntity file) {
		MetadataEntity meta = new MetadataEntity();
		meta.name = file.name;
		meta.type = file.type;
//		meta.parent = file.parent;  //TODO:
		meta.uri = file.uri;
		meta.documentId = file.documentId;
		meta.timestamp = file.timestamp;
		return meta;
	}

	@PrimaryKey(autoGenerate = true)
	public Long id;

	public String name;

	public int type;

	public Boolean processed;

	public String uri;

	// Unique documentId, we don't want duplicates from different root permissions
	public String documentId;

	public Long parentId;

	public String rating;

	public String label;

	public String timestamp;

	public String make;

	// Make and Model tables, only need model, make is implicit http://en.tekstenuitleg.net/articles/software/database-design-tutorial/second-normal-form.html
	public String model;

	public String aperture;

	public String exposure;

	public String flash;

	public String focalLength;

	public String iso;

	public String whiteBalance;

	public int height;

	public int width;

	public String latitude;

	public String longitude;

	public String altitude;

	public int orientation;

	public String lens;//todo: table

	public String driveMode;//todo: table

	public String exposureMode;//todo: table

	public String exposureProgram;//todo: table
}

File: app/src/main/java/com/anthonymandra/rawdroid/data/MetadataEntity.kt
package com.anthonymandra.rawdroid.data

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.ForeignKey.CASCADE
import androidx.room.Index
import androidx.room.PrimaryKey
import kotlinx.android.parcel.Parcelize

@Entity(tableName = "meta",
    foreignKeys = [ForeignKey(
        entity = FolderEntity::class,
        parentColumns = ["id"],
        childColumns = ["parentId"],
        onDelete = CASCADE)],
    indices = [
        Index(value = ["uri"], unique = true),
        Index(value = ["documentId"], unique = true),
        Index(value = ["parentId"])])
open class MetadataEntity(
    @PrimaryKey(autoGenerate = true)
    var id: Long = 0,
    var name: String = "",
    var type: Int = 0,
    var size: Long = 0,
    var processed: Boolean = false,
    var uri: String = "",
    var documentId: String = "",
    var parentId: Long = -1,
    var rating: Float? = null,
    var label: String? = null,   //TODO: This should be a table with nickname red: delete, etc.
    var timestamp: Long? = null,
    var make: String? = null,
    // Make and Model tables only need model make is implicit http://en.tekstenuitleg.net/articles/software/database-design-tutorial/second-normal-form.html
    var model: String? = null,
    var aperture: String? = null,
    var exposure: String? = null,
    var flash: String? = null,
    var focalLength: String? = null,
    var iso: String? = null,
    var whiteBalance: String? = null,
    var height: Int = 0,
    var width: Int = 0,
    var latitude: String? = null,
    var longitude: String? = null,
    var altitude: String? = null,
    var orientation: Int = 0,
    var lens: String? = null,//todo: table
    var driveMode: String? = null,//todo: table
    var exposureMode: String? = null,//todo: table
    var exposureProgram: String? = null//todo: table
) {
    // Simple equals regarding only gallery items
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as MetadataEntity

        if (id != other.id) return false
        if (name != other.name) return false
        if (uri != other.uri) return false
        if (rating != other.rating) return false
        if (label != other.label) return false
        if (timestamp != other.timestamp) return false

        return true
    }

    override fun toString(): String {
        return "MetadataEntity($documentId)"
    }
}


